File: apps/ios/App/Charts/GearRibbon.swift
import SwiftUI
import TrackEvolutionKit

/// The gear ribbon (#187) — the counterpart of `gearRibbonSvg` in
/// `public/js/gears.js`.
///
/// One horizontal band per highlighted lap on the channel panel's shared
/// driven-distance axis, cut into blocks of one gear with the number written in
/// any block wide enough to hold it. **Not a line chart**: `gear` is an enum, and
/// a line between 3 and 4 implies a gear no car has — which is also why the
/// importer samples it by holding the last value. Gear 0 (clutch in / no gear)
/// renders as a *gap* rather than a block, because it is genuinely "no gear".
///
/// With two or more laps lit, the runs where they sit in different gears are
/// outlined, and that is the whole feature: "you took T5 in 3rd on your best lap
/// and 4th on this one". The cutting and the comparison rule are the Kit's
/// `Gears`, pinned to the web implementation by `contracts/logic/gears.json`;
/// this file draws.
///
/// Hand-rolled on `Canvas` rather than Swift Charts: a run of one gear is a
/// filled rectangle with a label inside it, which is a bar chart of nothing, and
/// the panel already pays for one `Chart` per channel.
struct GearRibbon: View {
    let channels: SessionChannels
    /// Channel-lap indexes in slot order, as `LapChannelPanel` keeps them.
    let lit: [Int]
    let slots: [Color]
    /// The session's own lap number for a channel entry.
    let lapNumber: (Int) -> Int

    private struct Band {
        let chIdx: Int
        let color: Color
        let gear: [Double]
    }

    private var bands: [Band] {
        lit.enumerated().compactMap { slot, chIdx in
            guard slot < slots.count, channels.laps.indices.contains(chIdx),
                  let gear = channels.laps[chIdx].gear, !gear.isEmpty
            else { return nil }
            return Band(chIdx: chIdx, color: slots[slot], gear: gear)
        }
    }

    /// The same x-extent the channel charts use, so the ribbon lines up with the
    /// RPM trace above it: the longest lap's speed series, falling back to gear.
    private var span: Double {
        var maxN = 0
        for l in channels.laps {
            let n = l.speed?.count ?? l.gear?.count ?? 0
            if n > maxN { maxN = n }
        }
        return max(1, Double(maxN - 1) * channels.dStepM)
    }

    private static let rowHeight: CGFloat = 22
    private static let rowGap: CGFloat = 6

    var body: some View {
        let rows = bands
        if !rows.isEmpty {
            let disagreements = Gears.gearDisagreements(rows.map { $0.gear })
            VStack(alignment: .leading, spacing: 4) {
                HStack(alignment: .firstTextBaseline) {
                    Text("Gear")
                        .teStyle(.eyebrow)
                        .foregroundStyle(Color(.textMuted))
                    Spacer()
                    if rows.count >= 2 {
                        Text(disagreements.isEmpty ? "same gears throughout" : "dashed: laps disagree")
                            .teStyle(.xxs)
                            .foregroundStyle(Color(.textFaint))
                    }
                }
                HStack(alignment: .top, spacing: 8) {
                    VStack(alignment: .trailing, spacing: Self.rowGap) {
                        ForEach(rows, id: \.chIdx) { row in
                            Text("L\(String(lapNumber(row.chIdx)))")
                                .teStyle(.xxs)
                                .foregroundStyle(row.color)
                                .frame(height: Self.rowHeight)
                        }
                    }
                    Canvas { context, size in
                        draw(rows, disagreements, in: context, size: size)
                    }
                    .frame(height: CGFloat(rows.count) * Self.rowHeight + CGFloat(rows.count - 1) * Self.rowGap)
                }
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel("Gear by driven distance, per lap")
            .accessibilityValue(summary(rows, disagreements))
            .accessibilityIdentifier("gearRibbon")
        }
    }

    private func draw(
        _ rows: [Band], _ disagreements: [Gears.Disagreement], in context: GraphicsContext, size: CGSize
    ) {
        let step = channels.dStepM
        func x(_ distance: Double) -> CGFloat { CGFloat(distance / span) * size.width }

        for (index, row) in rows.enumerated() {
            let y = CGFloat(index) * (Self.rowHeight + Self.rowGap)
            let last = row.gear.count - 1
            for segment in Gears.gearSegments(row.gear) where segment.gear > 0 {
                let xa = x(max(0, Double(segment.k0) - 0.5) * step) + 1
                let xb = x(min(Double(last), Double(segment.k1) + 0.5) * step) - 1
                guard xb > xa else { continue }
                let rect = CGRect(x: xa, y: y, width: xb - xa, height: Self.rowHeight)
                context.fill(Path(roundedRect: rect, cornerRadius: 2), with: .color(row.color))
                // Only where the block can hold the digit — a number clipped to a
                // sliver reads as noise on the ribbon.
                if rect.width >= 14 {
                    let text = Text(String(Int(segment.gear)))
                        .font(.system(size: 11, weight: .bold))
                        .foregroundColor(Color(.surfaceCard))
                    context.draw(context.resolve(text), at: CGPoint(x: rect.midX, y: rect.midY))
                }
            }
        }

        guard rows.count >= 2 else { return }
        for run in disagreements {
            let xa = x(max(0, Double(run.k0) - 0.5) * step)
            let xb = x((Double(run.k1) + 0.5) * step)
            let rect = CGRect(x: xa, y: -2, width: xb - xa, height: size.height + 4)
            context.stroke(
                Path(roundedRect: rect, cornerRadius: 3),
                with: .color(Color(.danger)),
                style: .init(lineWidth: 1.5, dash: [3, 2])
            )
        }
    }

    /// What VoiceOver gets: the gears each lap used and where they disagree —
    /// the outline is precisely what a screen-reader user cannot see.
    private func summary(_ rows: [Band], _ disagreements: [Gears.Disagreement]) -> String {
        var parts = rows.map { row -> String in
            let used = Gears.gearSegments(row.gear)
                .filter { $0.gear > 0 }
                .map { Int($0.gear) }
            let low = used.min() ?? 0
            let high = used.max() ?? 0
            return "Lap \(String(lapNumber(row.chIdx))) used \(Gears.ordinal(Double(low))) to \(Gears.ordinal(Double(high)))"
        }
        if rows.count >= 2 {
            parts.append(
                disagreements.isEmpty
                    ? "The laps agree on gear throughout"
                    : "The laps take different gears in \(disagreements.count) place\(disagreements.count == 1 ? "" : "s")"
            )
        }
        return parts.joined(separator: ". ")
    }
}

File: apps/android/app/src/main/kotlin/app/trackevolution/ui/charts/GearRibbon.kt
package app.trackevolution.ui.charts

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.dp
import app.trackevolution.core.Gears
import app.trackevolution.core.model.SessionChannels
import app.trackevolution.ui.theme.TrackTheme
import kotlin.math.max
import kotlin.math.min

/**
 * The gear ribbon (#187) — the counterpart of `gearRibbonSvg` in
 * `public/js/gears.js`.
 *
 * One horizontal band per highlighted lap on the channel panel's shared
 * driven-distance axis, cut into blocks of one gear with the number written in
 * any block wide enough to hold it. **Not a line chart**: `gear` is an enum, and
 * a line between 3 and 4 implies a gear no car has — which is also why the
 * importer samples it by holding the last value. Gear 0 (clutch in / no gear)
 * renders as a *gap* rather than a block, because it is genuinely "no gear".
 *
 * With two or more laps lit, the runs where they sit in different gears are
 * outlined, and that is the whole feature: "you took T5 in 3rd on your best lap
 * and 4th on this one". The cutting and the comparison rule are `Gears` in
 * `:core`, pinned to the web implementation by `contracts/logic/gears.json`;
 * this file draws.
 */
@Composable
fun GearRibbon(
    channels: SessionChannels,
    /** Channel-lap indexes in slot order, as [LapChannelChart] keeps them. */
    lit: List<Int>,
    slots: List<Color>,
    /** The session's own lap number for a channel entry. */
    lapNumber: (Int) -> Int,
    modifier: Modifier = Modifier,
) {
    val rows = lit.mapIndexedNotNull { slot, chIdx ->
        val gear = channels.laps.getOrNull(chIdx)?.gear
        if (gear.isNullOrEmpty()) null else Triple(chIdx, slots[slot % slots.size], gear)
    }
    if (rows.isEmpty()) return

    val colors = TrackTheme.colors
    val density = LocalDensity.current
    val measurer = rememberTextMeasurer()
    val blockStyle = TrackTheme.typography.xxs.copy(color = colors.surfaceCard)

    val disagreements = Gears.gearDisagreements(rows.map { it.third })
    // The same x-extent the channel charts use, so the ribbon lines up with the
    // RPM trace above it: the longest lap's speed series, falling back to gear.
    val span = max(
        1.0,
        (channels.laps.maxOfOrNull { (it.speed?.size ?: it.gear?.size ?: 0) }?.minus(1) ?: 0) * channels.dStepM,
    )

    val summary = buildString {
        rows.forEachIndexed { i, (chIdx, _, gear) ->
            if (i > 0) append(". ")
            val used = Gears.gearSegments(gear).filter { it.gear > 0 }.map { it.gear.toInt() }
            append("Lap ${lapNumber(chIdx)} used ")
            append(Gears.ordinal((used.minOrNull() ?: 0).toDouble()))
            append(" to ${Gears.ordinal((used.maxOrNull() ?: 0).toDouble())}")
        }
        if (rows.size >= 2) {
            append(
                if (disagreements.isEmpty()) {
                    ". The laps agree on gear throughout"
                } else {
                    ". The laps take different gears in ${disagreements.size} " +
                        "place${if (disagreements.size == 1) "" else "s"}"
                },
            )
        }
    }

    val rowHeight = 22.dp
    val rowGap = 6.dp
    val height = rowHeight * rows.size + rowGap * (rows.size - 1)

    Column(
        modifier
            .fillMaxWidth()
            .clearAndSetSemantics {
                testTag = "gearRibbon"
                contentDescription = summary
            },
    ) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Text("Gear", style = TrackTheme.typography.xxs, color = colors.textMuted)
            if (rows.size >= 2) {
                Text(
                    if (disagreements.isEmpty()) "  same gears throughout" else "  dashed: laps disagree",
                    style = TrackTheme.typography.xxs,
                    color = colors.textFaint,
                )
            }
        }
        Row(Modifier.fillMaxWidth().padding(top = 2.dp), horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            Column(verticalArrangement = Arrangement.spacedBy(rowGap)) {
                rows.forEach { (chIdx, color, _) ->
                    Box(Modifier.height(rowHeight), contentAlignment = Alignment.CenterEnd) {
                        Text("L${lapNumber(chIdx)}", style = TrackTheme.typography.xxs, color = color)
                    }
                }
            }
            Canvas(Modifier.fillMaxWidth().height(height)) {
                val rowPx = with(density) { rowHeight.toPx() }
                val gapPx = with(density) { rowGap.toPx() }

                fun x(distance: Double) = (distance / span).toFloat() * size.width

                rows.forEachIndexed { index, (_, color, gear) ->
                    val y = index * (rowPx + gapPx)
                    val last = gear.size - 1
                    for (segment in Gears.gearSegments(gear)) {
                        if (segment.gear <= 0) continue // no gear: a gap, not a block
                        val xa = x(max(0.0, segment.k0 - 0.5) * channels.dStepM) + 1f
                        val xb = x(min(last.toDouble(), segment.k1 + 0.5) * channels.dStepM) - 1f
                        if (xb <= xa) continue
                        drawRoundRect(
                            color = color,
                            topLeft = Offset(xa, y),
                            size = Size(xb - xa, rowPx),
                            cornerRadius = CornerRadius(2f, 2f),
                        )
                        // Only where the block can hold the digit — a number
                        // clipped to a sliver reads as noise on the ribbon.
                        val text = measurer.measure(segment.gear.toInt().toString(), blockStyle)
                        if (xb - xa >= text.size.width + 6f) {
                            drawText(
                                text,
                                topLeft = Offset(
                                    (xa + xb) / 2f - text.size.width / 2f,
                                    y + rowPx / 2f - text.size.height / 2f,
                                ),
                            )
                        }
                    }
                }

                if (rows.size < 2) return@Canvas
                for (run in disagreements) {
                    val xa = x(max(0.0, run.k0 - 0.5) * channels.dStepM)
                    val xb = x((run.k1 + 0.5) * channels.dStepM)
                    drawRoundRect(
                        color = colors.danger,
                        topLeft = Offset(xa, -2f),
                        size = Size(xb - xa, size.height + 4f),
                        cornerRadius = CornerRadius(3f, 3f),
                        style = Stroke(
                            width = 1.5f * density.density,
                            pathEffect = PathEffect.dashPathEffect(
                                floatArrayOf(3f * density.density, 2f * density.density),
                            ),
                        ),
                    )
                }
            }
        }
    }
}
